### PR TITLE
vpp: init at 23.06

### PR DESCRIPTION
**Pull request by cariandrum22**
## Description of changes

VPP: Vector Packet Processing is a software network platform that provides a high-speed data plane, implementing Linux as a router or switch. 
See here for a description of the features offered: https://s3-docs.fd.io/vpp/23.06/aboutvpp/featurelist.html.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/vpp/quicly/default.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    make picotls-core picotls-openssl -j $NIX_BUILD_CORES -C build-picotls
+  '';
+
+  installPhase = ''
+    make -C build-quicly install
+    make picotls-core picotls-openssl -C build-picotls install
+  '';
+
+  meta = with lib; {
+    description = "";
```
Comment (yu-re-ka) [inline pkgs/servers/vpp/quicly/default.nix:53]:
```suggestion
    description = "An IETF QUIC protocol implementation, written from the ground up to be used primarily within the H2O HTTP server";
```

File: pkgs/servers/vpp/default.nix [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      vpp-api/vapi/vapi_json_parser.py \
+      --replace "#!/usr/bin/env python3" "#!$(command -v python)"
+
+    cmake $cmakeFlags .
+  '';
+
+  meta = with lib; {
+    homepage = "https://fd.io/";
+    description =
+      "VPP is a fast, scalable layer 2-4 multi-platform network stack";
```
Comment (yu-re-ka) [inline pkgs/servers/vpp/default.nix:124]: Descriptions should not contain the package name, as they will be displayed next to each other

```suggestion
      "A fast, scalable layer 2-4 multi-platform network stack";
```

File: pkgs/servers/vpp/default.nix [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  patches = [
+    ./0001-fix-nsh-plugin-loading.patch
+    ./0002-explicity-include-std-array.patch
+  ];
+
+  sourceRoot = "vpp/src";
+
+  quicly = callPackage ./quicly { };
+
+  nativeBuildInputs = [ llvmPackages_14.clang cmake dpdk' intel-ipsec-mb' xdp-tools' libbpf libmnl elfutils zlib quicly rdma-core' python310 python310Packages.ply openssl libuuid subunit pkg-config libpcap jansson libnl libdaq srtp' mbedtls_2 check ];
```
Comment (yu-re-ka) [inline pkgs/servers/vpp/default.nix:85]: Shouldn't `llvmPackages_14.clang` already be available, since `llvmPackages_14.stdenv` is used?

File: pkgs/servers/vpp/default.nix [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  };
+
+  patches = [
+    ./0001-fix-nsh-plugin-loading.patch
+    ./0002-explicity-include-std-array.patch
+  ];
+
+  sourceRoot = "vpp/src";
+
+  quicly = callPackage ./quicly { };
```
Comment (yu-re-ka) [inline pkgs/servers/vpp/default.nix:83]: Since quicly also provides a cli application, and could theoretically be used by other software, please add it to all-packages and take it as input of vpp like any other package.

File: nixos/modules/services/networking/vpp.nix [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+    ${cpuSection}
+  '' + optionalString (cfg.extraConfig != "") ''
+    ${cfg.extraConfig}
+  ''));
+in
+{
+
+  ###### interface
+
+  options = {
```
Comment (yu-re-ka) [inline nixos/modules/services/networking/vpp.nix:163]: I would prefer to use [RFC 0042](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md)-style structured settings option, instead of mirroring each config option of VPP inside the NixOS module. This is because a lot of individual NixOS option impact evaluation performance, and also have to be maintained with new VPP versions.

**Comment by yu-re-ka** [comment]
According to the recently implemented [RFC 0140](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md) you could also place the package files in `pkgs/by-name/vp/vpp/package.nix` and `pkgs/by-name/qu/quicly/package.nix`, and then they are automatically added to the package set ;)

**Comment by yu-re-ka** [comment]
Feel free to delegate me the task of implementing any of the changes suggested. I have a strong interest in getting this functionality into NixOS, and we are very thankful for your work on this.

File: nixos/modules/module-list.nix [hunk 6]
```diff
@@ -1091,6 +1091,7 @@
   ./services/networking/v2ray.nix
   ./services/networking/v2raya.nix
   ./services/networking/vdirsyncer.nix
+  ./services/networking/vpp.nix
```
Comment (AndersonTorres) [inline nixos/modules/module-list.nix:1094]: NixOS modules go in a separate commit.

File: nixos/tests/all-tests.nix [hunk 7]
```diff
@@ -857,6 +857,7 @@ in {
   victoriametrics = handleTest ./victoriametrics.nix {};
   vikunja = handleTest ./vikunja.nix {};
   virtualbox = handleTestOn ["x86_64-linux"] ./virtualbox.nix {};
+  vpp = handleTestOn ["x86_64-linux"] ./vpp.nix {};
```
Comment (AndersonTorres) [inline nixos/tests/all-tests.nix:860]: NixOS tests go in a separate commit

File: nixos/modules/services/networking/vpp.nix [hunk 8]
```diff
@@ -0,0 +1,658 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
```
Comment (AndersonTorres) [inline nixos/modules/services/networking/vpp.nix:3]:
```suggestion
```

File: nixos/modules/services/networking/vpp.nix [hunk 9]
```diff
@@ -0,0 +1,658 @@
...
+
+let
+  cfg = config.services.vpp;
+  vpp = pkgs.vpp;
```
Comment (AndersonTorres) [inline nixos/modules/services/networking/vpp.nix:7]: the package should be user-configurable

File: pkgs/servers/vpp/default.nix [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+, libnl
+, libdaq
+, srtp
+, mbedtls_2
+, check
+, nixosTests
+}:
+
+let
+  stdenv = llvmPackages_14.stdenv;
```
Comment (AndersonTorres) [inline pkgs/servers/vpp/default.nix:32]:
```suggestion
```

File: pkgs/servers/vpp/default.nix [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+  srtp' = srtp.overrideAttrs (old: rec {
+    mesonFlags = old.mesonFlags ++ [
+      "-Ddefault_library=static"
+    ];
+  });
+in
+stdenv.mkDerivation rec {
+  pname = "vpp";
+  version = "23.06";
+  src = fetchgit {
```
Comment (AndersonTorres) [inline pkgs/servers/vpp/default.nix:70]:
```suggestion

  src = fetchgit {
```

File: pkgs/servers/vpp/default.nix [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+  hardeningDisable = [ "fortify" "bindnow" ];
+
+  cmakeFlags = [
+    "-DCMAKE_BUILD_TYPE=release"
+    "-DCMAKE_INSTALL_PREFIX=${placeholder "out"}"
+    "-DVPP_USE_SYSTEM_DPDK=ON"
+  ];
+
+  configurePhase = ''
```
Comment (AndersonTorres) [inline pkgs/servers/vpp/default.nix:95]: This should be a `postPatch`, obviating the need of calling cmake at the end.

File: pkgs/servers/vpp/default.nix [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+      "plugins tools/vppapigen tools/g2 tools/perftool cmake"
+
+     # Replace hard-coded python with one that can be referenced from
+     # the built environment
+    substituteInPlace \
+      tools/vppapigen/vppapigen \
+      vpp-api/vapi/vapi_c_gen.py \
+      vpp-api/vapi/vapi_cpp_gen.py \
+      vpp-api/vapi/vapi_json_parser.py \
+      --replace "#!/usr/bin/env python3" "#!$(command -v python)"
```
Comment (AndersonTorres) [inline pkgs/servers/vpp/default.nix:116]: use patchShebangs instead

File: pkgs/servers/vpp/default.nix [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+      Some VPP Use-cases include vSwitches, vRouters, Gateways, Firewalls and
+      Load-Balancers, to name a few.
+    '';
+    license = with licenses; [ asl20 ];
+    maintainers = with maintainers; [ cariandrum22 ];
+    platforms = platforms.unix;
+  };
+
+  passthru.tests = { inherit (nixosTests) vpp; };
```
Comment (AndersonTorres) [inline pkgs/servers/vpp/default.nix:151]: By convention, passthru comes right before meta

File: pkgs/servers/vpp/quicly/0001-cmake-install.patch [hunk 15]
```diff
@@ -0,0 +1,30 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 5142bac..7b95f74 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -33,6 +33,8 @@ ENDIF(NOT CMAKE_BUILD_TYPE)
+ SET(CMAKE_C_FLAGS "-std=c99 -Wall -g ${CC_WARNING_FLAGS} ${CMAKE_C_FLAGS}")
+ SET(CMAKE_C_FLAGS_DEBUG "-O0")
+ SET(CMAKE_C_FLAGS_RELEASE "-O2")
++SET(CMAKE_POSITION_INDEPENDENT_CODE ON)
```
Comment (AndersonTorres) [inline pkgs/servers/vpp/quicly/0001-cmake-install.patch:9]: 1. It  should be upstreamed
2. Use [GNUInstallDirs](https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html)

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by cariandrum22** [comment]
@yu-re-ka @AndersonTorres Thank you for your review and comments. I would like to make fixes. Is there any guideline regarding the granularity of the fix commit?

**Comment by AndersonTorres** [comment]
Coarse. Very coarse. As coarse as you can.

Since the package is completely new and was not merged yet, you should introduce the fixups as they never existed to begin with.

**Comment by vifino** [comment]
Wow, impressive!

I packaged VPP for myself a long while ago for some projects but didn't bother upstreaming it (sorry).
While presumably far from generic, you can take a look and check if there is anything you might want to implement here (perhaps the buffer/hugepages math and vpp_papi packaging):

https://github.com/vifino/nix-geht/blob/master/pkgs/vpp/default.nix
https://github.com/vifino/nix-geht/blob/master/modules/vpp/default.nix

I didn't map all options but only added options for what made sense to me and instead chose to a dumb `extraConfig`, which is clearly different than what you are doing here.

When (or if) this gets merged, I'll switch to the derivation at least.
I also volunteer as co-maintainer.

File: pkgs/servers/vpp/default.nix (hunk 10 above)
Comment (vifino) [inline pkgs/servers/vpp/default.nix:32]: Why are you explicitly using LLVM?

File: pkgs/servers/vpp/default.nix [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+    version = "22.07";
+    src = fetchurl {
+      url = "https://fast.dpdk.org/rel/dpdk-${version}.tar.xz";
+      sha256 = "sha256-n2Tf3gdf21cIy2Leg4uP+4kVdf7R4dKusma6yj38m+o";
+    };
+  });
+
+  rdma-core' = rdma-core.overrideAttrs (old: rec {
+    cmakeFlags = old.cmakeFlags ++ [
+      "-DENABLE_STATIC=1"
```
Comment (vifino) [inline pkgs/servers/vpp/default.nix:57]: No need to build the shared libraries.
```suggestion
      "-DENABLE_STATIC=1"
      "-DBUILD_SHARED_LIBS:BOOL=false"
```

File: pkgs/servers/vpp/default.nix [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+  stdenv = llvmPackages_14.stdenv;
+
+  xdp-tools' = xdp-tools.overrideAttrs (old: rec {
+    postInstall = ''
+      # Drop unfortunate references to glibc.dev/include at least from $lib
+      nuke-refs "$lib"/lib/bpf/*.o
+    '';
+  });
+
+  intel-ipsec-mb' = intel-ipsec-mb.overrideAttrs (old: rec {
```
Comment (vifino) [inline pkgs/servers/vpp/default.nix:41]:
```suggestion
  intel-ipsec-mb' = intel-ipsec-mb.overrideAttrs (old: {
```

File: pkgs/servers/vpp/default.nix [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+
+  dpdk' = dpdk.overrideAttrs (old: rec {
+    version = "22.07";
+    src = fetchurl {
+      url = "https://fast.dpdk.org/rel/dpdk-${version}.tar.xz";
+      sha256 = "sha256-n2Tf3gdf21cIy2Leg4uP+4kVdf7R4dKusma6yj38m+o";
+    };
+  });
+
+  rdma-core' = rdma-core.overrideAttrs (old: rec {
```
Comment (vifino) [inline pkgs/servers/vpp/default.nix:55]:
```suggestion
  rdma-core' = rdma-core.overrideAttrs (old: {
```

File: pkgs/servers/vpp/default.nix [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+, srtp
+, mbedtls_2
+, check
+, nixosTests
+}:
+
+let
+  stdenv = llvmPackages_14.stdenv;
+
+  xdp-tools' = xdp-tools.overrideAttrs (old: rec {
```
Comment (vifino) [inline pkgs/servers/vpp/default.nix:34]:
```suggestion
  xdp-tools' = xdp-tools.overrideAttrs (old: {
```

File: pkgs/servers/vpp/default.nix [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+    };
+  });
+
+  rdma-core' = rdma-core.overrideAttrs (old: rec {
+    cmakeFlags = old.cmakeFlags ++ [
+      "-DENABLE_STATIC=1"
+    ];
+  });
+
+  srtp' = srtp.overrideAttrs (old: rec {
```
Comment (vifino) [inline pkgs/servers/vpp/default.nix:61]:
```suggestion
  srtp' = srtp.overrideAttrs (old: {
```

**Review by vifino — CHANGES_REQUESTED** [review]
Just a quick skimming of the vpp derivation.

I think you should also add options to make the DPDK/rdma-core/srtp plugins optional, as they are not always required or desired.
For example, if you have Mellanox cards, they work with both DPDK/rdma-core.

If you're just using virtio/tap devices, you might not want either.

Since VPP is modular, maybe packaging them individually (after building them together? I don't recall how free-standing the build process for them is) is possible, but that's probably non-trivial.
I'd love a `plugins = with pkgs.vppPlugins; [ dpdk linux_nl linux_cp ];`, though.

**Comment by wfdewith** [comment]
@cariandrum22 I'm pretty interested in this; is there any chance that you will continue working on this PR?
